apis/v1beta1/validation/gateway.go

```diff
@@ -118,3 +121,39 @@ func ValidateTLSCertificateRefs(listeners []gatewayv1b1.Listener, path *field.Pa
 	}
 	return errs
 }
+
+// ValidateListenerNames validates the names of the listeners
+// must be unique within the Gateway
+func ValidateListenerNames(listeners []gatewayv1b1.Listener, path *field.Path) field.ErrorList {
+	var errs field.ErrorList
+	nameMap := map[gatewayv1b1.SectionName]struct{}{}
+	for i, c := range listeners {
+		if _, found := nameMap[c.Name]; found {
+			errs = append(errs, field.Forbidden(path.Index(i).Child("name"), fmt.Sprintln("must be unique within the Gateway")))
+		}
+		nameMap[c.Name] = struct{}{}
+	}
+	return errs
+}
+
+// validateHostnameProtocolPort validates that the combination of port, protocol, and hostname are
+// unique for each listener.
+func validateHostnameProtocolPort(listeners []gatewayv1b1.Listener, path *field.Path) field.ErrorList {
+	var errs field.ErrorList
+	hostnameProtocolPortSets := sets.Set[string]{}
+	for i, listener := range listeners {
+		hostname := new(gatewayv1b1.Hostname)
+		if listener.Hostname != nil {
+			hostname = listener.Hostname
+		}
+		protocol := listener.Protocol
+		port := listener.Port
+		hostnameProtocolPort := fmt.Sprintf("%s:%s:%d", *hostname, protocol, port)
+		if hostnameProtocolPortSets.Has(hostnameProtocolPort) {
```

Comment:
I think the k8s set util that we're using here is roughly doing the same thing but has a nicer interface to work with (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/sets/set.go#L24). Anything I'm missing?

Reply:
I found some places still use `map` instead of k8s set, such as https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1beta1/validation/gateway.go#L129, https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1beta1/validation/httproute.go#L110. I think we can unify them.